include/oneapi/dpl/pstl/histogram_binhash_utils.h

```diff
@@ -0,0 +1,123 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_HISTOGRAM_BINHASH_UTILS_H
+#define _ONEDPL_HISTOGRAM_BINHASH_UTILS_H
+#include <algorithm>
+#include <utility>
+#include <iterator>
+#include <cstdint>
+#include <type_traits>
+#include <cassert>
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace __internal
+{
+template <typename _T1, typename = void>
+struct __evenly_divided_binhash;
+
+template <typename _T1>
+struct __evenly_divided_binhash<_T1, ::std::enable_if_t<::std::is_floating_point_v<_T1>>>
+{
+    _T1 __minimum;
+    _T1 __maximum;
+    _T1 __scale;
+
+    __evenly_divided_binhash(const _T1& __min, const _T1& __max, ::std::size_t __num_bins)
+        : __minimum(__min), __maximum(__max), __scale(_T1(__num_bins) / (__max - __min))
+    {
+        assert(__num_bins < ::std::numeric_limits<::std::int32_t>::max());
+    }
+
+    template <typename _T2>
+    ::std::int32_t
+    get_bin(_T2 __value) const
+    {
+        ::std::int32_t ret = -1;
+        if ((__value >= __minimum) && (__value < __maximum))
+        {
+            ret = (__value - __minimum) * __scale;
+        }
+        return ret;
+    }
+};
+
+template <typename _T1>
+struct __evenly_divided_binhash<_T1, ::std::enable_if_t<!::std::is_floating_point_v<_T1>>>
+{
+    _T1 __minimum;
+    _T1 __range_size;
+    ::std::int32_t __num_bins;
+    __evenly_divided_binhash(const _T1& __min, const _T1& __max, ::std::size_t __num_bins_)
+        : __minimum(__min), __num_bins(__num_bins_), __range_size(__max - __min)
+    {
+        assert(__num_bins < ::std::numeric_limits<::std::int32_t>::max());
+    }
+
+    template <typename _T2>
+    ::std::int32_t
+    get_bin(_T2 __value) const
+    {
+        ::std::int32_t ret = -1;
+        if ((__value >= __minimum) && (__value < (__minimum + __range_size)))
+        {
+            ret = ((::std::uint64_t(__value) - __minimum) * ::std::uint64_t(__num_bins)) / __range_size;
```

Comment:
The first operand `((::std::uint64_t(__value) - __minimum)` is correct?
Probably should be  `::std::uint64_t(__value - __minimum)` ?

Reply:
Yes, fixed.  I've also improved the testing a bit to try to have covered that case.  However, I think that the code was underflowing uint64 in a way to end up "accidentally" at the correct result anyway after the subtraction.